Jint/Engine.cs

```diff
@@ -849,14 +860,19 @@ public void DeclarationBindingInstantiation(DeclarationBindingType declarationBi
             }
 
             // process all variable declarations in the current parser scope
-            foreach (var d in variableDeclarations.SelectMany(x => x.Declarations))
+            for (var i = 0; i < variableDeclarations.Count; i++)
             {
-                var dn = d.Id.As<Identifier>().Name;
-                var varAlreadyDeclared = env.HasBinding(dn);
-                if (!varAlreadyDeclared)
+                var variableDeclaration = variableDeclarations[i];
+                var declarations  = (List<VariableDeclarator>) variableDeclaration.Declarations;
```

Comment:
this is a repeating pattern, we know Esprima and that it produces List<T>, we can cast to it can get non-virtual dispatch + allocation free foreach

Reply:
Might be worth just changing it there, no? That would give also skip the type checks here.

Reply:
I have a pending pull request https://github.com/sebastienros/esprima-dotnet/pull/13 that optimizes things, but does not change the API. Same thing here about setting the public API to concrete implementations. I guess streaming API like IEnumerable will not be supported in the near future and everything is materialized as is.

Forced cast is pretty much safe though as unit tests cover well all these and upgrading to different kind of Esprima would show up.

Reply:
I've created separate pull request to change Ast to have concerete types: https://github.com/sebastienros/esprima-dotnet/pull/14

Reply:
Concrete type PR went in, but requires NuGet release to be used here.

Reply:
Pushed on Nuget, should appear soon.